tests/unit/test_signers.py

```diff
@@ -401,6 +469,49 @@ def test_generate_presigned_post(self):
                          'https://s3.amazonaws.com/mybucket')
         self.assertEqual(post_form_args['fields'], {})
 
+    def test_generate_presigned_post_emits_choose_signer(self):
+        with mock.patch.dict(botocore.auth.AUTH_TYPE_MAPS,
+                             {'s3v4-presign-post': self.auth}):
+            self.signer.generate_presigned_post(self.request_dict)
+        self.emitter.emit_until_response.assert_called_with(
+            'choose-signer.service_name.PutObject',
+            signing_name='signing_name', region_name='region_name',
+            signature_version='s3v4-presign-post')
+
+    def test_generate_presigned_post_choose_signer_override(self):
+        auth = mock.Mock()
+        self.emitter.emit_until_response.return_value = (None, 'custom')
+        auth_types = {
+            's3v4-presign-post': self.auth,
+            'custom-presign-post': auth
+        }
+        with mock.patch.dict(botocore.auth.AUTH_TYPE_MAPS, auth_types):
+            self.signer.generate_presigned_post(self.request_dict)
+        auth.assert_called_with(
+            credentials=self.fixed_credentials, region_name='region_name',
+            service_name='signing_name')
+
+    def test_generate_presigned_post_bad_signer_raises_error(self):
```

Comment:
You covered the not happy case. I was thinking what if the user returned s3v4-presign-post for the presign_post, I would expect it would properly do the presign post without any issues. You covered the case of mismatching for presign posts and urls.